deployer.go

```diff
@@ -165,25 +167,125 @@ func (d *deployer) destroyUnwanted(wantedUnits, currentUnits map[string]*schema.
 	return nil
 }
 
-func (d *deployer) launchAll(wantedUnits, currentUnits map[string]*schema.Unit) error {
+func (d *deployer) launchAll(wantedUnits, currentUnits map[string]*schema.Unit, serviceCount map[string]int) error {
+	deployedUnits := make(map[string]bool)
+
 	for _, u := range wantedUnits {
+		// unit may have already been deployed - all nodes of a service get deployed,
+		// when the first one appears in the wanted list
+		if _, ok := deployedUnits[u.Name]; ok {
+			continue
+		}
+
 		if u.DesiredState == "" {
 			u.DesiredState = "launched"
 		}
+
+		//units that are changed are set to Inactive in deployUnit()
 		if currentUnits[u.Name].DesiredState != u.DesiredState {
-			err := d.fleetapi.SetUnitTargetState(u.Name, u.DesiredState)
+
+			if !needsSequentialDeployment(u.Name, serviceCount) {
+				err := d.fleetapi.SetUnitTargetState(u.Name, u.DesiredState)
+				if err != nil {
+					return err
+				}
+				continue
+			}
+
+			deployed, err := d.performSequentialDeployment(u, serviceCount)
 			if err != nil {
 				return err
 			}
+			for k, v := range deployed {
+				deployedUnits[k] = v
+			}
 		}
 	}
+
 	return nil
 }
 
+func (d *deployer) performSequentialDeployment(u *schema.Unit, serviceCount map[string]int) (map[string]bool, error) {
+	deployedUnits := make(map[string]bool)
+	serviceName := strings.Split(u.Name, "@")[0]
+	nrOfNodes := serviceCount[serviceName]
+
+	for i := 1; i <= nrOfNodes; i++ {
+		// generate unit name with number - the one we have originally might not be
+		// the 1st node
+		unitName := fmt.Sprintf("%v@%d.%v", serviceName, i, strings.Split(u.Name, ".")[1])
+
+		// start the service
+		err := d.fleetapi.SetUnitTargetState(unitName, u.DesiredState)
+		if err != nil {
+			return nil, err
+		}
+
+		sidekickName := strings.Replace(unitName, "@", "-sidekick@", 1)
+
+		//record the fact we deployed these
+		deployedUnits[unitName] = true
+		deployedUnits[sidekickName] = true
+
+		//if it's the last one of the cluster, we're done
+		if i == nrOfNodes {
+			break
+		}
+
+		// every second, check if the corresponding sidekick is up
+		// we do this for a maximum of 1 minute
+		timeoutChan := make(chan bool)
+		go func() {
+			<-time.After(time.Duration(5) * time.Minute)
```

Comment:
Is 5 mins too long?  Also, it doesn't match the comment.  You can use a time.Timer for this functionality